### PR TITLE
Add `deleteRecord(for key:)` to `ReadWriteTransaction`

### PR DESCRIPTION
**Pull request by designatednerd**
This comes with a ton of caveats, but I tried to document them in both code and in the docs. 

Note that we're going to be looking at a much more robust system for cache eviction and possibly garbage collection in the future, but this is a stopgap until that's ready to go live.

### CODE REVIEW

**Comment by designatednerd** [comment]
Addresses #1555

**Comment by designatednerd** [comment]
Gonna let @martijnwalraven take a look at this before I merge it to make sure I'm not doing anything _too_ dumb here

File: Sources/Apollo/ApolloStore.swift [hunk 1]
```diff
@@ -241,6 +241,10 @@ public final class ApolloStore {
       try body(&object)
       try write(object: object, withKey: key, variables: variables)
     }
+    
+    public func removeRecord(for key: CacheKey) throws {
```
Comment (martijnwalraven) [inline Sources/Apollo/ApolloStore.swift:245]: For consistency with the other methods, I would make this `removeObject(withKey:)`. (I don't think we should use record as part of the public API in general, only for cache implementors.)

File: Sources/ApolloSQLite/SQLiteNormalizedCache.swift [hunk 2]
```diff
@@ -127,6 +132,10 @@ extension SQLiteNormalizedCache: NormalizedCache {
     return try mergeRecords(records: records)
   }
   
+  public func removeRecord(for key: CacheKey) throws {
+    try removeSQLiteRecord(for: key)
```
Comment (martijnwalraven) [inline Sources/ApolloSQLite/SQLiteNormalizedCache.swift:136]: Why the separate `removeSQLiteRecord()`? Wouldn't `removeRecord()` on the SQLite cache already imply that?

File: docs/source/caching.mdx [hunk 3]
```diff
@@ -177,3 +177,38 @@ store.withinReadWriteTransaction({ transaction in
   }
 })
 ```
+
+### delete
+
+Delete functionality is limited at this time. There are presently two options available on `NormalizedCache` which are supported by the included implementations of it:
```
Comment (martijnwalraven) [inline docs/source/caching.mdx:183]: > There are presently two options available on `NormalizedCache` which are supported by the included implementations of it
This sentence seems a bit hard to understand, and I would avoid `NormalizedCache`. Maybe something like:
> There are presently two deletion methods available on `ApolloStore`, which are supported by both the in memory and the SQLite cache.

File: docs/source/caching.mdx [hunk 4]
```diff
@@ -177,3 +177,38 @@ store.withinReadWriteTransaction({ transaction in
...
 ```
+
+### delete
+
+Delete functionality is limited at this time. There are presently two options available on `NormalizedCache` which are supported by the included implementations of it: 
+
+1. `clear` - Removes everything from the cache immediately. This is basically the "Nuke it from orbit, it's the only way to be sure" option. 
+2. `removeRecord(for:)` Removes a single record for the given `CacheKey`. 
+
+`removeRecord(for:)` will only remove things at the level of a record - that is, it cannot remove individual fields from a record, and it cannot remove outside references **to** a record.
```
Comment (martijnwalraven) [inline docs/source/caching.mdx:188]: I wouldn't use the term record here, but talk about removing objects because records are not part of the public API.

File: docs/source/caching.mdx [hunk 5]
```diff
@@ -177,3 +177,38 @@ store.withinReadWriteTransaction({ transaction in
...
+
+Delete functionality is limited at this time. There are presently two options available on `NormalizedCache` which are supported by the included implementations of it: 
+
+1. `clear` - Removes everything from the cache immediately. This is basically the "Nuke it from orbit, it's the only way to be sure" option. 
+2. `removeRecord(for:)` Removes a single record for the given `CacheKey`. 
+
+`removeRecord(for:)` will only remove things at the level of a record - that is, it cannot remove individual fields from a record, and it cannot remove outside references **to** a record. 
+
+```swift
+// Assuming we start with records of the following format:
```
Comment (martijnwalraven) [inline docs/source/caching.mdx:191]: I think in order to be understandable, this needs a much more detailed explanation of cache normalization and the structure of records. It may make sense to move the implementation details to a separate section and focus on the API and observable behavior here.

**Review by martijnwalraven — CHANGES_REQUESTED** [review]

File: Sources/Apollo/ApolloStore.swift (hunk 1 above)
Comment (designatednerd) [reply Sources/Apollo/ApolloStore.swift:245]: Fair

File: Sources/ApolloSQLite/SQLiteNormalizedCache.swift (hunk 2 above)
Comment (designatednerd) [reply Sources/ApolloSQLite/SQLiteNormalizedCache.swift:136]: to match the other things in this file that have a wrapper method

File: docs/source/caching.mdx (hunk 5 above)
Comment (designatednerd) [reply docs/source/caching.mdx:191]: Gonna work on this bit tomorrow when I have a couple functioning brain cells

File: docs/source/caching.mdx [hunk 6]
```diff
@@ -177,3 +177,31 @@ store.withinReadWriteTransaction({ transaction in
...
+
+```swift
+apollo.cacheKeyForObject = { $0["id"] }
+```
+
+This would indicate that for every object which has an `id` property, it will be cached with that `id` as the key. This would be ideal for an API which uses globally unique identifiers, as our Star Wars API does. 
+
+This means that if you have previously fetched an object with the ID `"2001"`, you can call `transaction.removeObject(for: "2001")` on a `ReadWriteTransaction` to remove any object cached with that key, along with all of its associated scalar properties and the references to other objects it stores. 
+
+`removeObject` does _not_ cascade its removals - if you remove an object which has reference to another object, the reference will be removed, but that other object will not be removed and will remain in the cache. Likewise, if you delete an object, references _to_ that object will not be deleted, they will simply fail when you attempt to load the object again.
```
Comment (martijnwalraven) [inline docs/source/caching.mdx:202]: Maybe mention that depending on the cache policy, failure because of dangling references doesn't necessary imply an error but will just result in a cache miss and refetching the data from the server.

**Review by martijnwalraven — APPROVED** [review]

File: docs/source/caching.mdx (hunk 6 above)
Comment (designatednerd) [reply docs/source/caching.mdx:202]: yeah I think I can rephrase this to "to fail, causing a cache miss," which should theoretically clarify that a bit more.
